cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go

```diff
@@ -211,6 +213,22 @@ func (c *NamespacedResourceWatcherCache) isGvrValid() error {
 	return fmt.Errorf("CRD [%s] is not valid", c.config.Gvr)
 }
 
+func (c *NamespacedResourceWatcherCache) syncAndStartWatchLoop(stopCh <-chan struct{}) {
+	// RetryWatcher will take care of re-starting the watcher if the underlying channel
+	// happens to close for some reason, as well as recover from other failures
+	// at the same time ensuring not to replay events that have been processed
+	watcher, err := c.resyncAndNewRetryWatcher(true)
+	if err != nil {
+		err = fmt.Errorf(
+			"[%s]: Initial resync failed after [%d] retries were exhausted, last error: %v",
+			c.queue.Name(), maxWatcherCacheRetries, err)
+		// yes, I really want this to panic. Something is seriously wrong and
+		// possibly restarting kubeapps-apis server is needed...
```

Comment:
Is it possible that we'll panic here simply because Redis isn't yet ready (that's what was happening when I created this issue - Redis seems to take a while to come up on my local cluster, which was causing kubeapps-apis to fail to be ready, prior to this change). With this change, I assume kubeapps-apis will become ready faster, but this code will still cause the restart if redis isn't ready in time? Not sure how we could handle it... perhaps explicitly handling waiting for redis to become ready initially, in this go-routine, but before we do the resync with panic?

Reply:
Well, I could increase maxWatcherCacheResyncBackoff from 2 to say 8). That way we will wait a LONG time before giving up

Reply:
Ah great. I missed the ping.

> We may or may not want to change that behavior, e.g. by introducing retries w/ exponential back off in that routine (`NewRedisClientFromEnv`) or maybe some kind of loop to wait for redis to come up. I'd like to see some evidence first that things are broken (log files will suffice)

Yep, I'd be keen (eventually, if/when you agree) to do a loop waiting for redis to come up with some timeout, so it only affects the initial load. As for evidence, what I see whenever I enable flux is first this:

```
k -n kubeapps get po
NAME                                              READY   STATUS             RESTARTS   AGE
kubeapps-c49cb4c5d-vfxmf                          2/2     Running            0          20s
kubeapps-internal-dashboard-5b77d89ff-hxf4g       1/1     Running            0          20s
kubeapps-internal-kubeappsapis-5465c7b6d4-mzmjh   1/1     Running            0          2d23h
kubeapps-internal-kubeappsapis-c8c597f5f-jw5rs    0/1     CrashLoopBackOff   1          20s
kubeapps-internal-kubeops-7646db7767-xd6tx        1/1     Running            0          20s
kubeapps-redis-master-0                           0/1     Running            0          20s
kubeapps-redis-replicas-0                         0/1     Running            0          20s
```

So here redis is still coming up, so kubeappsapis has already restarted once. The logs show:

```
k -n kubeapps logs kubeapps-internal-kubeappsapis-c8c597f5f-jw5rs
I0307 05:13:06.075910       1 root.go:37] kubeapps-apis has been configured with: core.ServeOptions{Port:50051, PluginDirs:[]string{"/plugins/fluxv2", "/plugins/resources"}, ClustersConfigPath:"/config/clusters.conf", PluginConfigPath:"/config/kubeapps-apis/plugins.conf", PinnipedProxyURL:"http://kubeapps-internal-pinniped-proxy.kubeapps:3333", GlobalReposNamespace:"kubeapps", UnsafeLocalDevKubeconfig:false, QPS:50, Burst:100}
I0307 05:13:06.671936       1 main.go:22] +fluxv2 RegisterWithGRPCServer
I0307 05:13:06.671984       1 server.go:60] +fluxv2 NewServer(kubeappsCluster: [default], pluginConfigPath: [/config/kubeapps-apis/plugins.conf]
Error: failed to initialize plugins server: failed to register plugins: plug-in "name:\"fluxv2.packages\" version:\"v1alpha1\"" failed to register due to: dial tcp 10.96.16.162:6379: connect: connection refused
...
```

Once redis is ready, it settles, but is always left with a number of restarts (3 for me), which makes it look like there's been some unexpected problem:

```
k -n kubeapps get po                                             
NAME                                             READY   STATUS    RESTARTS   AGE
kubeapps-c49cb4c5d-vfxmf                         2/2     Running   0          2m23s
kubeapps-internal-dashboard-5b77d89ff-hxf4g      1/1     Running   0          2m23s
kubeapps-internal-kubeappsapis-c8c597f5f-jw5rs   1/1     Running   3          2m23s
kubeapps-internal-kubeops-7646db7767-xd6tx       1/1     Running   0          2m23s
kubeapps-redis-master-0                          1/1     Running   0          2m23s
kubeapps-redis-replicas-0                        1/1     Running   0          2m23s
```

Reply:
Wow, that looks pretty convincing.  Thanks. Wonder why I have not come across this myself with all the testing I've done?

I will point out that if this is the case, then `resync()` issue was a red herring. Meaning, this was what was causing the CrashLoopBackOff, not the indexing of the bitnami repo on startup

 Anyway, I will introduce a similar loop with exponential back off into `NewRedisClientFromEnv` to work around this

Reply:
> Wow, that looks pretty convincing. Thanks. Wonder why I have not come across this myself with all the testing I've done?

Maybe you don't start without Redis each time? (so it's already running). In my case I'm switching between flux, carvel and helm support (for demoing).

> 
> I will point out that if this is the case, then `resync()` issue was a red herring. Meaning, this was what was causing the CrashLoopBackOff, not the indexing of the bitnami repo on startup

Two separate issues I think. Those 3 restarts are due to redis not being ready, but then once Redis is ready, the plugin would take 30s to sync before the plugin itself completed its registration. Now because the server itself doesn't start serving until all plugins have been registered successfully, the readiness check would continue to fail during that time, which meant more restarts (I saw some where I had 6 or 7 restarts, but not everytime).

> 
> Anyway, I will introduce a similar loop with exponential back off into `NewRedisClientFromEnv` to work around this

Great, thanks, though in this case, we may not want exponential backoff, given that we want to be able to start with minimal delay - maybe just keep pinging ever second or two or something. See what you think.

Reply:
no problem. I will try ping a total of say 10 times, every second then give up?